src/mono/wasm/runtime/storage.ts

```diff
@@ -0,0 +1,170 @@
+    // caches will be undefined if we're running on an insecure origin (secure means https or localhost)
+    if (typeof globalThis.caches === "undefined") {
+        return null;
+    }
+
+    // cache integrity is compromised if the first request has been served over http (except localhost)
+    // in this case, we want to disable caching and integrity validation
+    if (ENVIRONMENT_IS_WEB && globalThis.window.isSecureContext === false) {
+        return null;
+    }
+
+    // Define a separate cache for each base href, so we're isolated from any other
+    // Blazor application running on the same origin. We need this so that we're free
+    // to purge from the cache anything we're not using and don't let it keep growing,
+    // since we don't want to be worst offenders for space usage.
+    const relativeBaseHref = document.baseURI.substring(document.location.origin.length);
+    const cacheName = `dotnet-resources${relativeBaseHref}`;
+
+    try {
+        // There's a Chromium bug we need to be aware of here: the CacheStorage APIs say that when
+        // caches.open(name) returns a promise that succeeds, the value is meant to be a Cache instance.
+        // However, if the browser was launched with a --user-data-dir param that's "too long" in some sense,
+        // then even through the promise resolves as success, the value given is `undefined`.
+        // See https://stackoverflow.com/a/46626574 and https://bugs.chromium.org/p/chromium/issues/detail?id=1054541
+        // If we see this happening, return "null" to mean "proceed without caching".
+        return (await globalThis.caches.open(cacheName)) || null;
+    } catch {
+        // There's no known scenario where we should get an exception here, but considering the
+        // Chromium bug above, let's tolerate it and treat as "proceed without caching".
+        console.warn("MONO_WASM: Failed to open cache");
+        return null;
+    }
+}
+
+export async function getMemorySnapshotSize(): Promise<number | undefined> {
+    try {
+        const inputsHash = await getInputsHash();
+        if (!inputsHash) {
+            return undefined;
+        }
+        const cacheKey = `${memoryPrefix}-${ProductVersion}-${GitHash}-${inputsHash}`;
+        const cache = await openCache();
+        if (!cache) {
+            return undefined;
+        }
+        const res = await cache.match(cacheKey);
+        const contentLength = res?.headers.get("content-length");
+        return contentLength ? parseInt(contentLength) : undefined;
+    } catch (ex) {
+        console.warn("MONO_WASM: Failed find memory snapshot in the cache", ex);
+        return undefined;
+    }
+}
+
+export async function getMemorySnapshot(): Promise<ArrayBuffer | undefined> {
+    try {
+        const inputsHash = await getInputsHash();
+        if (!inputsHash) {
+            return undefined;
+        }
+        const cacheKey = `${memoryPrefix}-${ProductVersion}-${GitHash}-${inputsHash}`;
```

Comment:
why do we duplicate the creation of `cacheKey`, `openCache` call, and `cache.match` in these two functions? Is it expected that we will call `getMemorySnapshotSize` and then `getMemorySnapshot` sequentially in a single run?
I have mild concern about the duplication allowing bugs to creep in, but also that we could end up paying some costs to load the snapshot from cache twice (once to get the size, then to actually get it) since they are separate response objects.

Reply:
moved into `getCacheKey`